File: logFiles.js
const input = [
  ["58523", "user_1", "resource_1"],
  ["62314", "user_2", "resource_2"],
  ["54001", "user_1", "resource_3"],
  ["200", "user_6", "resource_5"],
  ["215", "user_6", "resource_4"],
  ["54060", "user_2", "resource_3"],
  ["53760", "user_3", "resource_3"],
  ["58522", "user_4", "resource_1"],
  ["53651", "user_5", "resource_3"],
  ["2", "user_6", "resource_1"],
  ["100", "user_6", "resource_6"],
  ["400", "user_7", "resource_2"],
  ["100", "user_8", "resource_2"],
  ["54359", "user_1", "resource_3"],
  ]
  
function checkLogs(logList) {
  const userTimestamps = {};
  for (let i = 0; i < logList.length; i++) {
    let log = logList[i];
    if (!(log[1] in userTimestamps)) {
      userTimestamps[log[1]] = [parseInt(log[0]), parseInt(log[0])];
    } else {
      if (parseInt(log[0]) < userTimestamps[log[1]][0]) {
        userTimestamps[log[1]][0] = parseInt(log[0])
      } else if (parseInt(log[0]) > userTimestamps[log[1]][1]) {
        userTimestamps[log[1]][1] = parseInt(log[0]);
      }
    }
  }
  return userTimestamps;
}

function accessTimes(logList) {
let logs = logList
  .map(data => {
    const [time, ...rest] = data;
    return [parseInt(time), ...rest]
  }).sort((a, b) => a[0] - b[0]);
  let pointer = 0, frequency = {}, maxInstances = Number.NEGATIVE_INFINITY, maxResource = '';
  for (let i = 0; i < logs.length; i++) {
    const [timestamp, , resource] = logs[i];
    let [timestampTemp, , resourceTemp] = logs[pointer];
    let timeDiff = timestamp - timestampTemp;
    let isMoreThan5Min = timeDiff > 300;
    frequency[resource] = (frequency[resource] || 0) + 1;
    if (frequency[resource] > maxInstances && !isMoreThan5Min) {
      maxInstances = frequency[resource];
      maxResource = resource;
    }
    // console.log({i, pointer, frequency, maxInstances, timestampTemp, timestamp})
    while (isMoreThan5Min) {
      frequency[resourceTemp]--;
      pointer++;
      [timestampTemp, , resourceTemp] = logs[pointer];
      timeDiff = timestamp - timestampTemp;
      isMoreThan5Min = timeDiff > 300;
    }       
  }
  return [maxResource, maxInstances];
}

function probabilities(log_list) {
  // sort logs by time
  let logs = log_list
  .map(data => {
    const [time, ...rest] = data;
    return [parseInt(time), ...rest]
  }).sort((a, b) => a[0] - b[0]);
  const usersToResources = {};
  // build obj showing the resource(s) logged by each user
  for (let i = 0; i < logs.length; i++) {
    const [ , user, resource] = logs[i];
    if (!(usersToResources[user])) { 
      usersToResources[user] = ["__START__"];
    }
    usersToResources[user].push(resource);
  }
  const resourcesToNext = {};
  for (user in usersToResources) {
    const resources = usersToResources[user];
    resources.push("__END__");
    // build obj showing how many times a particular resource (destination) follows a given resource (origin)
    for (let i = 0; i < resources.length - 1; i++) {
      const origin = resources[i], destination = resources[i + 1];
      // the value of this obj is a map 
      if (!(resourcesToNext[origin])) {
        resourcesToNext[origin] = {};
      } 
      const resourceMap = resourcesToNext[origin];
      // add to the count of the destination resource every time it follows another resource
      resourceMap[destination] = (resourceMap[destination] || 0) + 1;
    }
  };
  // convert numbers to probabilities
  for (resource in resourcesToNext) {
    const nextResources = resourcesToNext[resource];
    // count how many times any resource follows the key resouce
    const total = Object.values(nextResources).reduce((a, b) => a + b, 0);
    for (nextResource in nextResources) {
      // rest from integer to probability
      nextResources[nextResource] = nextResources[nextResource] / total;
    }
  }
  return resourcesToNext;
}

console.log(checkLogs(input));
console.log(accessTimes(input));
console.log(probabilities(input));

File: log_files.py
logs = [
    ["58523", "user_1", "resource_1"],
    ["62314", "user_2", "resource_2"],
    ["54001", "user_1", "resource_3"],
    ["215", "user_6", "resource_4"],
    ["200", "user_6", "resource_5"],    
    ["54060", "user_2", "resource_3"],
    ["53941", "user_3", "resource_3"],
    ["58522", "user_4", "resource_1"],
    ["53651", "user_5", "resource_3"],
    ["2", "user_6", "resource_3"],
    ["100", "user_6", "resource_6"],
    ["400", "user_7", "resource_2"],
    ["100", "user_8", "resource_2"],
]

def sortLogs(logs):
    usersToTimes = {}
    for time, userId, resourceId in logs:
        time = int(time)
        if userId not in usersToTimes:
            usersToTimes[userId] = [time, time]
        else:
            if usersToTimes[userId][0] > time:
                usersToTimes[userId][0] = time
            
            if usersToTimes[userId][1] < time:
                usersToTimes[userId][1] = time
        
    return usersToTimes

# print(sortLogs(logs))

def mostUsedResource(logs):
    logs = sorted(logs, key=lambda x: int(x[0]))
    
    resourcesToTimes = {}
    
    for time, _, resource in logs:
        if resource not in resourcesToTimes:
            resourcesToTimes[resource] = []
        resourcesToTimes[resource].append(int(time))
        
    maxCount = 0
    maxResource = None
    for resource in resourcesToTimes.keys():
        # check first and see how many are in window of 300
        times = resourcesToTimes[resource]
        for idx in range(0, len(times)):
            initialTime = times[idx]
            
            for checkIdx in range(idx, len(times)):
                checkTime = times[checkIdx]
                if checkTime > initialTime + 300:
                    checkTime -= 1
                    break

            count = checkIdx - idx + 1
            if count > maxCount:
                maxCount = count
                maxResource = resource


    return maxResource, maxCount
    
# print(mostUsedResource(logs))


def probabilities(logs):
    logs = sorted(logs, key=lambda x: int(x[0]))
    
    usersToOrderedResources = {}
    
    for _, user, resource in logs:
        if user not in usersToOrderedResources:
            usersToOrderedResources[user] = ['__START__']
        usersToOrderedResources[user].append(resource)
        
    resourcesToNext = {}
    
    for user in usersToOrderedResources:
        resources = usersToOrderedResources[user]
        resources.append('__END__')
        
        for idx in range(0, len(resources) - 1):
            origin = resources[idx]
            destination = resources[idx + 1]
            if origin not in resourcesToNext:
                resourcesToNext[origin] = {}
                
            resourceMap = resourcesToNext[origin]
            if destination not in resourceMap:
                resourceMap[destination] = 0
            resourceMap[destination] += 1
        
    
    for resource in resourcesToNext:
        nexts = resourcesToNext[resource]
        
        total = sum(nexts.values())
        
        for nextResource in nexts:
            nexts[nextResource] = nexts[nextResource] * 1.0 / total
            
    return resourcesToNext
    
print(probabilities(logs))